ledger/eval/eval.go

```diff
@@ -1407,6 +1390,35 @@ func (eval *BlockEvaluator) endOfBlock() error {
 			return fmt.Errorf("fees collected wrong: %v != %v", eval.block.FeesCollected, expectedFeesCollected)
 		}
 
+		// agreement will check that the proposer is correct (we can't because
+		// we don't see the bundle), but agreement allows the proposer to be set
+		// even if Mining is not enabled (and unset any time).  So make sure
+		// it's set iff it should be.
+		if eval.proto.Mining().Enabled {
+			if eval.block.Proposer.IsZero() && !eval.generate { // if generating, proposer is set later by agreement
+				return fmt.Errorf("proposer missing when mining enabled")
+			}
+		} else {
+			if !eval.block.Proposer.IsZero() {
+				return fmt.Errorf("proposer %v present when mining disabled", eval.block.Proposer)
+			}
+		}
+
+		// agreement will check that the payout is zero if the proposer is
+		// ineligible, but we must check that it is correct if non-zero. We allow it
+		// to be too low. A proposer can be algruistic.
```

Comment:
I thought this was a new dictionary word at first.

Reply:
Puns are allowed.